### ISSUE TITLE
Feature request: Add support for Selenium grid 

### ISSUE THREAD

**Issue by jawlitkp** [anshooarora/extentreports#22]
Hi Anshoo,
Requesting you to add GRID support!

Let me know i am willing to test beta version

Thanks 
Jwalit

**Comment by anshooarora** [comment]
Jawlit - 

To be able to write to a single file from multiple machines is not a good approach.  I will have to create a semaphore that handles this:  http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Semaphore.html and it "should" work.  However, my worry is the word should.

A safer way to do this would be to:

* Create multiple XML files and join them into one
* Use a database and combine db info into one

I prefer the 1st one because then, there will be no dependency on 3rd party jars.  All xml code can be handled using native java packages and we can delete all XML files at the end if not required.  This feature is quite huge so it might take some time also.

This will also be beneficial to create test trends - how has testing evolved over a period of time.  By parsing each XML file, you will be able to know how many times a test has past over a period, how many tests have run, what has been an average pass rate etc.

Let me know your thoughts.

**Comment by jawlitkp** [comment]
Thanks Anshoo, as long as I can capture results i am good with your suggestion,Database would be more complicated so i would say XML approach is better one.

**Comment by virenv** [comment]
I think this should be solved at the Grid usage level. Extent report has already abstracted out a ExtentTest class. Which means that you can get your ExtentTest for each test and pass it to a machine in the Grid. Let the logging happen on ExtentTest and once the test is done, you can always write some code to do ExtentReports.endTest(ExtentTest) in a synchronized way.

I think having synchronized methods on ReportInstance should take care of adding ExtentTest data from multiple machines.

**Comment by jawlitkp** [comment]
Hello,

Actually at this what i am doing is i have one main server machine from node(appium servers) are on multiple machines but you are right my main machine is sending out commands to appium servers and getting response back so actual logging is done on main machine,allow me few more days to investigate further ,i will get back to you with results soon,you can close this ticket for now.

Again thanks for helping

**Comment by virenv** [comment]
Lets wait for your investigations. It will be great to see how we can solve this situation.

**Comment by aartinarang** [comment]
Hi,

I am facing similar problem here. I am using selenium grid for parallel execution on multiple devices at the same time. Its a multi-threaded operation, so my selenium server is sending requests to multiple nodes/devices simultaneously. So, I am quiet not sure if extent reports will be able to handle the same and give me multiple reports for different testng suites. Can you please help me with the same.

**Comment by anshooarora** [comment]
What version are you using?  Version 2 fully supports parallel runs for both classes and methods.  See here for example: http://extentreports.relevantcodes.com/2x/docs.html#parallel-classes

If you need multiple reports, you would have to create multiple extentReport instances and handle them all.

**Comment by anshooarora** [comment]
Also, please see the comment by Virender above for further clarification.  He has summed up your scenario with details very nicely: https://github.com/anshooarora/extentreports/issues/22#issuecomment-125423241

I am closing this ticket as this is a supported feature for parallelism and user has to decide the best approach to implement.  

If there is a specific request, feel free to open a new ticket.

**Comment by aartinarang** [comment]
Thanks for your reply Anshoo. My Bad, I missed those details. The link mentioned above, in your comment is really helpful and tells it all. I just have one confusion, the "file-path" mentioned here,
ExtentReports extent = new ExtentReports(file-path, replaceExisting);
is it where the report is generated?
I have implemented the same in my project, but I cannot find where my reports are getting generated?
I have used the above code-line as

ExtentReports extent = new ExtentReports("C:/AppAutomation/Report.html", false);

Thanks.
